File: JAVA-DSA/SecondLargestEleArray.java
/*

    Find the second largest element in a array

 */

import java.util.Arrays;

class SecondLargestEleArray {
    // brute foce solution
    // time complexity: O(NlogN + N)
    public static int solution1(int arr[]) {
        Arrays.sort(arr);
        int n = arr.length;
        int largest = arr[n-1];
        
        for ( int i = arr.length - 1 ; i >= 0 ; i-- ) {
            if ( arr[i] != largest ){
                return arr[i];
            }
        }

        return -1;
    }

    // better solution
    // time complexity: O(2N)
    public static int solution2(int arr[]) {
        int largest = arr[0];
        int Slargest = -1;
        int n = arr.length;
        for ( int i = 0 ; i < n ; i++ ) {
            if ( arr[i] > largest ) {
                largest = arr[i];
            }
        }

        for ( int i = 0 ; i < n ; i++ ) {
            if ( arr[i] > Slargest && arr[i] != largest ) {
                Slargest = arr[i];
            }
        }

        return Slargest;
    }

    // optimal solution
    // time complexity : O(N)
    public static int solution3(int arr[]) {
        int largest = arr[0];
        int slargest = -1;
        int n = arr.length;

        for ( int i = 0 ; i < n ; i++ ) {
            if ( arr[i] > largest ) {
                slargest = largest;
                largest = arr[i];
            } else if ( arr[i] < largest && arr[i] > slargest ) {
                slargest = arr[i];
            }
        }

        return slargest;
    }

    public static void main(String[] args){
        int arr[] = { 1, 2, 4, 7, 7, 5 };

        System.out.println("brute solution -> " + solution1(arr));
        System.out.println("better solution --> " + solution2(arr));
        System.out.println("optimal solution --> " + solution3(arr));
    }
}

File: Python-DSA/SecondLargestEleArray.py
"""

    find the second largest element in a array

"""

class SecondLargestEleArray:
    # brute force solution
    # time complexity: O(NlogN + N)
    @staticmethod
    def solution1(arr):
        arr.sort()
        largest = arr[-1]  

        for i in range(len(arr) - 1, -1, -1):
            if arr[i] != largest:
                return arr[i]

        return -1

    # better solution
    # time complexity: O(2N)
    @staticmethod
    def solution2(arr):
        largest = arr[0]
        slargest = -1

        for i in range(len(arr)):
            if arr[i] > largest:
                largest = arr[i]

        for i in range(len(arr)):
            if arr[i] > slargest and arr[i] != largest:
                slargest = arr[i]

        return slargest

    # optimal solution
    # time complexity: O(N)
    @staticmethod
    def solution3(arr):
        largest = arr[0]
        slargest = -1

        for i in range(len(arr)):
            if arr[i] > largest:
                slargest = largest
                largest = arr[i]
            elif arr[i] < largest and arr[i] > slargest:
                slargest = arr[i]

        return slargest

if __name__ == "__main__":
    arr = [1, 2, 4, 7, 7, 5]

    print("brute solution ->", SecondLargestEleArray.solution1(arr))
    print("better solution ->", SecondLargestEleArray.solution2(arr))
    print("optimal solution ->", SecondLargestEleArray.solution3(arr))
